dot/parachain/grid/grid_topology.go

```diff
@@ -0,0 +1,346 @@
+	_, y := gn.ValidatorIndicesCol[origin]
+
+	if x && y {
+		// If all works correctly origin peer can't be in both rows and columns. But we leave it anyways
+		return RequiredRoutingGridXY
+	}
+	if !(x || y) {
+		return RequiredRoutingNone
+	}
+	if x && !y {
+		return RequiredRoutingGridY
+	}
+	return RequiredRoutingGridX
+}
+
+// ShouldRouteToPeer indicates does peer should receive a message based on GridTopology and Routing strategy
+func (gn *GridNeighbors) ShouldRouteToPeer(routing RequiredRouting, peer peer.ID) bool {
+	switch routing {
+	case RequiredRoutingAll:
+		return true
+	case RequiredRoutingNone, PendingTopology:
+		return false
+	case RequiredRoutingGridXY:
+		_, x := gn.PeersRow[peer]
+		_, y := gn.PeersCol[peer]
+		return x || y
+	case RequiredRoutingGridX:
+		_, x := gn.PeersRow[peer]
+		return x
+	case RequiredRoutingGridY:
+		_, y := gn.PeersCol[peer]
+		return y
+	default:
+		// No way we get here
+		return false
+	}
+}
+
+// PeersDiff returns a differents between two GridNeighbors
+func (gn *GridNeighbors) PeersDiff(other *GridNeighbors) []peer.ID {
+	diff := make([]peer.ID, 0)
+	for p := range gn.PeersRow {
+		_, inRows := other.PeersRow[p]
+		_, inCols := other.PeersCol[p]
+		if !inRows && !inCols {
+			diff = append(diff, p)
+		}
+	}
+	return diff
+}
+
+func (gn *GridNeighbors) Len() int {
+	return len(gn.PeersRow) + len(gn.PeersCol)
+}
+
+type RequiredRouting uint
+
+const (
+	PendingTopology    RequiredRouting = iota
+	RequiredRoutingAll                 // All peers in the grid
+	RequiredRoutingGridXY
+	RequiredRoutingGridX
+	RequiredRoutingGridY
+	RequiredRoutingNone
+)
+
+// MatrixNeighbors holds the row and column neighbors of a given index in a matrix.
+type MatrixNeighbors struct {
+	RowNeighbors    []uint
+	ColumnNeighbors []uint
+}
+
+// CalculateMatrixNeighbors computes the row and column neighbors of valIndex in a matrix of given length.
+// e.g. for size 11 the matrix would be
+//
+// 0  1  2
+// 3  4  5
+// 6  7  8
+// 9 10
+//
+// and for index 10, the neighbors would be 1, 4, 7, 9
+func CalculateMatrixNeighbors(valIndex, length uint) (*MatrixNeighbors, error) {
+	if valIndex >= length {
+		return nil, fmt.Errorf("grid topology malformed: valIndex %d is greater than length %d",
+			valIndex,
+			length,
+		)
+	}
+
+	sqrt := uint(math.Sqrt(float64(length)))
+	ourRow := valIndex / sqrt
+	ourColumn := valIndex % sqrt
+
+	rowStart := ourRow * sqrt
+	rowEnd := uint(math.Min(float64(rowStart+sqrt), float64(length)))
+
+	rowNeighbors := make([]uint, 0)
+	for i := rowStart; i < rowEnd; i++ {
+		if i != valIndex {
+			rowNeighbors = append(rowNeighbors, i)
+		}
+	}
+
+	columnNeighbors := make([]uint, 0)
+	for i := ourColumn; i < length; i += sqrt {
+		if i != valIndex {
+			columnNeighbors = append(columnNeighbors, i)
+		}
+	}
+
+	return &MatrixNeighbors{
+		RowNeighbors:    rowNeighbors,
+		ColumnNeighbors: columnNeighbors,
+	}, nil
+}
+
+type SessionGridTopologyEntry struct {
```

Comment:
The types `SessionGridTopologyEntry` and `SessionGridTopologyStorage` is not part of the scope of Session Grid Topology. These types are related to how a given session grid is stored and that is specific to each subsystem that needs to store this info. 
- https://github.com/ChainSafe/gossamer/blob/development/docs/docs/design/grid-topology.md#session-grid-topology-storage

Reply:
I have decided to implement SessionGridTopologyStorage since it is used in statement distirbution and bitfield distribution. SessionGridTopologies which is not implemented here is a map way of storing topologies is used in Approval Ditribution so I decided not to implement it since Approval distribution is out of scope for Validator MVP. 

Nevertheless I do not see much problems having it implemented. When can move it then to those packages if needed or leave here.